src/t_hash.c

```diff
@@ -1051,11 +1050,57 @@ SetExRes hashTypeSetEx(redisDb *db, robj *o, sds field, HashTypeSet *setKeyVal,
     return res;
 }
 
+/*
+ * hashTypeSetExRdb provide a simplified API for setting fields & expiry by RDB load
+ *
+ * It is the duty of RDB reading process to track minimal expiration time of the
+ * fields and eventually call hashTypeAddToExpires() to update global HFE DS with
+ * next expiration time.
+ *
+ * To just add a field with no expiry, use hashTypeSet instead.
+ */
+int hashTypeSetExRdb(redisDb *db, robj *o, sds field, sds value, uint64_t expire_at) {
+    /* Dummy struct to be used in hashTypeSetEx() */
+    HashTypeSetEx setEx = {
+            .fieldSetCond = FIELD_DONT_OVRWRT,         /* Shouldn't be any duplication */
+            .expireSetCond = HFE_NX,                   /* Should set expiry once each field */
+            .minExpire = EB_EXPIRE_TIME_INVALID,       /* Won't be used. Accounting made by RDB already */
+            .key = NULL,                               /* Not going to call hashTypeSetExDone() */
+            .hashObj = o,
+            .minExpireFields = EB_EXPIRE_TIME_INVALID, /* Not needed by RDB */
+            .c = NULL,                                 /* No notification required */
+            .cmd = NULL,                               /* No notification required */
+    };
+
+    HashTypeSet setKeyVal = {.value = value, .flags = 0};
+    SetExRes res = hashTypeSetEx(db, o, field, &setKeyVal, expire_at, (expire_at) ? &setEx : NULL);
+    return (res == HSETEX_OK || res == HSET_UPDATE) ? C_OK : C_ERR;
+}
+
+void initDictExpireMetadata(sds key, robj *o) {
+    dict *ht = o->ptr;
+
+    dictExpireMetadata *m = (dictExpireMetadata *) dictMetadata(ht);
+    m->key = key;
+    m->hfe = ebCreate();     /* Allocate HFE DS */
+    m->expireMeta.trash = 1; /* mark as trash (as long it wasn't ebAdd()) */
+}
+
 /*
  * Init HashTypeSetEx struct before calling hashTypeSetEx()
  *
  * Don't have to provide client and "cmd". If provided, then notification once
  * done by function hashTypeSetExDone().
+ *
+ * NOTE: when calling from RDB reading process, the key is not yet available
+ * in the DB. Therefore, it is not possible to retrieve a pointer to the key to
+ * be used in the expire meta struct.
+ * The expireMeta struct needs to hold a pointer to the key string that is
+ * persistent for the key lifetime. When reading RDB, the key was already read
+ * and the string it was read into will be later used in the DB. However,
+ * if calling this function from any place else and using this flag, you'll need
+ * to provide a key string in the key robj that will remain persistent for as
+ * long as the key itself is.
```

Comment:
This comment has become a stale one right?

Reply:
right, removed